other-docs/guides/upgrading/v3.md

```diff
@@ -0,0 +1,143 @@
+
+Tests are run in your local environment, and provide full integration testing capabilities to match production, including S3, ElasticSearch, caching and Tachyon.
+
+This feature provides flexibility for more complex use cases, check out the [unit testing documentation and examples here](docs://dev-tools/testing-with-phpunit.md).
+
+
+### Browser Security
+
+Good security is a growing concern online, and the new [browser security](docs://security/browser.md) feature set provides a configurable solution for securing static assets and cross origin behaviour. Supported features include:
+
+- Automatic [Subresource Integrity](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) for first-party assets, and a framework for third-party assets
+- Automatic `X-XSS-Protection` and `X-Frame-Options` headers by default
+- Content Security Policy support, with smart defaults and highly configurable
+
+
+## Other Developer Features and APIs
+
+### Core Module Loading (Core Module)
+
+Modules were previously loaded via Composer's autoload functionality, however this approach was limiting, as the entrypoint files contained side-effects and required a `function_exists()` check at the start of the file to prevent non-Altis CLI commands from breaking.
+
+Modules now add an `extra.altis` section to their `composer.json` to be included a generated file called `vendor/modules.php`. This file is included from `wp-config.php`.
+
+The [advice for writing custom modules](docs://getting-started/custom-modules.md) has now been updated to reflect this but any existing custom modules will remain backwards compatible.
+
+
+### Developer Tools Additions (Developer Tools module)
+
+Query Monitor integration is now vastly improved with clickable traces for both PHP function calls and database queries that open in your editor.
+
+In addition X-Ray data and the flamegraph are now available as panels in Query Monitor.
+
+A [default Travis CI config](docs://dev-tools/continuous-integration.md) will be created if one does not already exist that will run the zero config PHPUnit tests.
+
+
+### Local Environment improvements (Local Chassis & Local Server modules)
+
+There is a new `composer chassis exec` command for running single commands directly on the VM and additionally [you can now extend the default chassis configuration from your `composer.json`](docs://local-chassis/) like so:
+
+```json
+{
+	"extra": {
+		"altis": {
+			"modules": {
+				"local-chassis": {
+					"hosts": [
+						"altis.local",
+						"subsite.altis.local"
+					],
+					"extensions": [
+						"xdebug"
+					]
+				}
+			}
+		}
+	}
+}
+```
+
+If you make any changes to your Local Chassis config in `composer.json` run `composer chassis provision` to apply them.
+
+As with the Local Chassis module there is now a `composer local-server exec` command for running arbitrary commands on the PHP docker container.
```

Comment:
Sorry, was imprecise before; let's combine this part with the `chassis exec` part, and just have something like:

> Both Local Server and Local Chassis now support an `exec` command for running arbitrary commands in the development environment. `composer local-chassis exec` runs a command inside the virtual machine, while `composer local-server exec` runs a command inside the PHP Docker container

Reply:
ah I see